[2단계 - 자동차 경주 리팩토링] 이오(이지우) 미션 제출합니다.

LJW25 opened the pull request:
안녕하세요 서브웨이, 이전 단계에서 남겨주셨던 리뷰를 참고하여 리팩토링해보았습니다!
가장 크게 달라진 부분은 전진 여부를 결정하는 난수 생성을 별도의 Interface로 분리하고, 이를 이용한 테스트를 구현한 것 입니다.

추가로 입력값 검증에 관하여 고민이 되는 부분이 있는데, 아래의 코멘트 확인해주시면 감사하겠습니다.

리뷰 너무 감사드리고, 좋은하루되세요 :)

LJW25 committed:
refactor(RacingCarController) : 함수 순서 변경

- 내려가기 규칙을 참고하여 함수 순서를 변경함.

LJW25 committed:
refactor(RacingCarController) : 불필요한 static 제거

LJW25 committed:
refactor(CarRandomNumberGenerator) : 클래스 이름 변경

LJW25 committed:
refactor(CarNumberGenerator) : 난수 생성 Interface 구현

LJW25 committed:
refactor(Car) : 메소드 간결화

LJW25 committed:
refactor(RacingCarController) : Generator 분리

LJW25 committed:
test(ApplicationTest) : 전체 테스트 추가

LJW25 committed:
test(ApplicationTest) : 전체 테스트 리팩터링

LJW25 committed:
test(ApplicationTest) : 전체 테스트 리팩터링

LJW25 commented:
차 이름 입력값 검증에 대해 몇가지 방향을 고민중에 있습니다. 조원들과도 이야기 나눠 보았지만, 명확하게 해결되지 않아 괜찮으시다면 서브웨이의 의견도 여쭤보고 싶어요!

처음 경기를 진행 할 차의 이름을 입력 받을 때, 제가 제한한 예외 조건은 크게 세가지 입니다.

- 빈 값인 경우
- 5글자를 초과한 경우
- 중복된 경우

현재 제가 구현한 코드는 모든 검증을 InputView에서 수행하고 있는데, Car의 특성에 따른 검증을 InputView에서 수행해도 되는가에 대한 고민을 계속해서 하고있습니다.

현재까지 생각한 방향은 크게 세가지인데요

1. 검증하는 메소드를 모두 InputView에서 수행하도록 구현한 경우: 검증이 Car의 특성에 종속되기 때문에, Car의 예외 조건이 변경 될 경우 InputView에 수정이 필요합니다. 
2. 차의 이름과 관련된 검증은 Car 내부에서 수행하도록 구현한 경우: 이 경우 각각의 Car에서 검증하기 때문에, 중복된 이름을 검사하기 어렵다는 문제가 있습니다.
3. 중복된 경우만 InputView에서 수행하고, 이외의 예외는 Car에서 검증하도록 구현한 경우: 최대한 도메인을 독립적으로 구성할 수 있도록 노력하였지만, 중복된 경우는 여전히 InputView에서 검증하고 있으므로 역할을 나뉘어가진 상태라 오히려 애매하다고 느껴졌습니다.

혹시 서브웨이는 이 중에 어떤 방향이 가장 적절하다고 생각하시나요? 혹은 추천하고싶은 다른 방향이 있으실까요?

제가 고민했던 내용을 잘 설명하고 싶은데, 깔끔하지 못한걸 보니 아직은 질문하는 것도 연습이 필요할것같네요 😂 
의견 남겨주시면 감사하겠습니다!

joseph415 commented:
안녕하세요 이오!

저의 생각을 말씀드리자면, 저는 두번째 방식이 맞다고 생각합니다!

하지만, 이오가 말씀주신것처럼 차의 이름과 관련된 검증은 Car 도메인에서 해야하는 검증이라고 생각해요! 그런데, 저희가 과제로 하고있는 자동차경주 특성상 2번으로 구현을 하기 쉽지 않을 것 같아요ㅎㅎ; (중복 이름을 검사하려면, 저장소에서 Car를 가지고 있어야하고 항상 추적이 가능한 상태여야 할텐데, db를 사용한다거나, 메모리 저장소를 직접 구현하는것은 오버엔지니어링이라고 생각합니다)

따라서, 3번의 방향으로 할 것 같아요. 1번의 방식은 이오가 생각한 것 처럼, 비지니스 검증을 View에서 하는것은 올바른 책임이 아니라고 생각합니다!

깊게 고민하는 부분이 인상깊네요ㅎㅎ 추가적으로 질문사항 있으시면, 커멘트 달아주세요~

joseph415 commented on src/main/java/racingcar/CarNumberGenerator.java:
```diff
@@ -0,0 +1,6 @@
+package racingcar;
+
+@FunctionalInterface
```
👍

joseph415 commented on src/main/java/racingcar/Application.java:
```diff
@@ -5,7 +5,8 @@
 
     public static void main(String[] args) {
-        RacingCarController racingCarController = new RacingCarController();
+        CarRandomNumberGenerator carRandomNumberGenerator = new CarRandomNumberGenerator();
```
CarRandomNumberGenerator 자동차 움직임 규칙이 들어간 중요한 비지니스 라고 생각하는데, 해당 위치가 맞을까요?

joseph415 commented on src/main/java/racingcar/Application.java:
```diff
@@ -5,7 +5,8 @@
 
     public static void main(String[] args) {
-        RacingCarController racingCarController = new RacingCarController();
+        CarRandomNumberGenerator carRandomNumberGenerator = new CarRandomNumberGenerator();
```
저는 도메인 모델은 외부로부터 보호받아야하는 순수한 객체라고 생각하는데요. 컨트롤러 외부에서 주입해주면, 외부에서 도메인 객체를 의존하게 되는 상황이 발생해서, 도메인이 외부로 노출된다고 생각합니다!

따라서 Controller에서 직접 의존해서 사용하는건 어떤지 의견드리고싶습니다!

https://github.com/woowacourse/java-racingcar/pull/477#discussion_r1102127205

콩하나와 나눴던 이야기인데요. 내용 한번 참고해보시고, 답변 부탁드릴게요! 
이오 의견은 어떤지 듣고싶습니다.

LJW25 replied:
남겨주신 리뷰 참고해서 오늘 하루 틈틈히 고민해보았습니다!

우선 CarRandomNumberGenerator는 주요한 도메인이기 때문에 컨트롤러 내부에 있어야 한다는 말씀은 동의합니다!(사실 동의보다는 이해라는 표현이 적절할것같습니다. 제가 미처 생각하지 못했던 부분이라서요ㅎㅎ)

제가 CarRandomNumberGenerator를 해당 위치에 넣어 줬던 이유는 제가 구현한 방식 상에서 난수생성은 RacingCarController의 tryOneTime 메소드 내에서 호출되었기 때문입니다!
해당 메소드는 private로 선언되어 있어 직접적으로 carNumberGenerator를 바꿔가며 테스트하는게 불가능했기 때문에, 컨트롤러를 생성할 때 입력값으로 주는 방식으로 변경 가능하도록 구현하고자 하였습니다.

말씀해 주신 것처럼 CarRandomNumberGenerator를 보호하는 동시에 private 메소드를 public으로 바꾸지 않고, 변경 가능성도 주기 위해서 tryOneTime 메소드를 RacingCar 클래스에서 수행하도록 옮기고, CarNumberGenerator를 인자로 넘겨주는 방식을 사용하였습니다. CarRandomNumberGenerator 선언은 컨트롤러에서 하도록 구현하였습니다.

이 방식을 사용할 경우 전체 ApplicationTest가 불가능하다는 단점이 있지만, 더 작은 단위의 테스트로 대체할 수 있고 분리가 가능하다는 장점이 더 크다고 생각하여 해당 방식으로 수정하였습니다.
혹시 더 나은 방법이 있다면 조언해주시면 감사하겠습니다 :)

joseph415 replied:
전체 테스트도 중요하지만, 저도 이오가 생각한것처럼 도메인 로직을 테스트하는 것이 더 중요하다고 생각합니다! 
결국에 보호해야할 영역은 그쪽이니까요ㅎㅎ.. 그 둘을 선택해야 한다면 저는 이오가 선택한 부분을 선택할 것 같아요!

joseph415 commented on src/main/java/racingcar/CarNumberGenerator.java:
```diff
@@ -0,0 +1,6 @@
+package racingcar;
+
+@FunctionalInterface
+public interface CarNumberGenerator {
```
CarNumberGenerator 테스트가 빠져있네요..!

LJW25 replied:
CarNumberGenerator의 경우 자바에서 제공하는 Random을 사용하여 난수를 생성 할 뿐이라 별도의 테스트를 구현하지 않았습니다! 혹시 이 부분에도 테스트가 필요할까요?
PR시 언급했던 테스트는 전체 ApplicationTest를 의미했었습니다. 이 테스트의 경우 carRandomNumberGenerator 관련 코드가 수정되면서 사용이 불가능해 졌기 때문에 삭제후 더 작은 단위의 테스트로 대체하였습니다.

joseph415 reviewed (changes requested):
안녕하세요 이오! 2단계 구현 잘해주셨네요 💯 

대체적으로 코드가 깔끔해졌네요~! 
추가적으로 몇가지 코멘트 드렸는데, 확인부탁드립니다!

LJW25 committed:
refactor(RacingCarController): 난수 생성 위치 변경

- 난수 생성을 호출하는 tryOneTime 메소드 클래스 변경(RacingCarController -> RacingCars)
- CarRandomNumberGenerator를 Application에서 사용하지 않도록 수정

LJW25 committed:
test(ApplicationTest): 전체 테스트 삭제

LJW25 committed:
test(RacingCarsTest): 한 회 진행 테스트 추가

LJW25 commented:
> 따라서, 3번의 방향으로 할 것 같아요. 1번의 방식은 이오가 생각한 것 처럼, 비지니스 검증을 View에서 하는것은 올바른 책임이 아니라고 생각합니다!

감사합니다!
이론적으로는 2번이 맞다고 생각하면서도 구현할 방법이 생각나지 않아 답답했었는데, 드디어 해소가 됬네요😄
최대한 책임을 분리하되, 지금처럼 작은 규모의 미션에서 불가피한 부분은 View에서 일부 대체하는 방식으로 생각하면 될 것 같습니다.

피드백 너무 감사드립니다! :)

joseph415 reviewed (approved):
안녕하세요 이오!
2단계 마지막 피드백까지 구현하느라 고생하셨습니다! 💯 
요구사항은 모두 잘 만족하신것 같아서 이만 merge 하겠습니다. 👍 
추가적으로 간단한 피드백 남겼습니다~!

joseph415 commented:
안녕하세요 이오! 2단계 피드백 적용해주신것 확인했습니다!

제가 요청드린것은 사실
1. 'CarNumberGenerator, CarRandomNumberGenerator 는 비지니스객체이기때문에 model 패키지로 이동해야한다.' 
2. CarRandomNumberGenerator가 비지니스 모델이기 때문에, 별도의 테스트가 필요하지 않을까? 
이였습니다.

2번을 테스트하기 위해서는 TestNumberGenerator 처럼 테스트용 객체를 만들어서 무조건 통과하는 테스트를 작성하게 될텐데요. 
테스트는 동작하는 문서 역할을 하기 때문에 이 또한 의미 있다고 생각해요ㅎㅎ 누군가가 해당 테스트를 보고 자동차의 규칙이 랜덤으로 이동하는 비지니스 로직을 파악할 테니까요.

하지만, 이오가 RacingCarTest에서 작성해주신 부분도 좋다고 생각하기 때문에, 이번단계는 merge 했습니다!
2단계 구현하면서, 사다리 같이 진행하시느라 고생많으셨어요ㅎㅎ
